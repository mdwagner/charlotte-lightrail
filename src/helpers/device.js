import { Dimensions } from 'react-native';

// identify iPhone model based on device width
const identifyDevice = (deviceScreenProps) => {
  if (deviceScreenProps.width === 414) {
    return 'iPhone 6+';
  } else if (deviceScreenProps.width === 320) {
    return 'iPhone 5';
  }
  return 'iPhone 6';
};

const deviceScreen = Dimensions.get('window');
const deviceName = identifyDevice(deviceScreen);
let deviceVariableSizes = {};
let defaultCenter = {};
let defaultZoom;
if (deviceName === 'iPhone 6+') {
  deviceVariableSizes = {
    calloutHeight: 300,
    nextTimeFontSize: 40,
    stationLabelFontSize: 36,
    nextTimeWrapPaddingVertical: 4,
    nearestStationText: 'Nearest Station - '
  };
  defaultCenter = {
    latitude: 35.12848262558094,
    longitude: -80.84703007335676
  };
  defaultZoom = 10.35016331854935;
} else if (deviceName === 'iPhone 6') {
  deviceVariableSizes = {
    calloutHeight: 279,
    nextTimeFontSize: 36,
    stationLabelFontSize: 30,
    nextTimeWrapPaddingVertical: 1,
    nearestStationText: 'Nearest Station - '
  };
  defaultCenter = {
    latitude: 35.09018630471958,
    longitude: -80.84307324707783
  };
  defaultZoom = 9.228717656576594;
} else {
  deviceVariableSizes = {
    calloutHeight: 279,
    nextTimeFontSize: 33,
    stationLabelFontSize: 22,
    nextTimeWrapPaddingVertical: 1,
    nearestStationText: 'Nearest - '
  };
  defaultCenter = {
    latitude: 35.12642689061146,
    longitude: -80.8549019571087
  };
  defaultZoom = 10.10400032344327;
}

const calloutBoxHeight = deviceVariableSizes.calloutHeight - 25; // excludes triangle
const blueBoxHeight = calloutBoxHeight * 0.45; // 45% of calloutBoxHeight
const blackBoxHeight = calloutBoxHeight * 0.55; // 55% of calloutBoxHeight

const insideColPadding = deviceScreen.width * 0.13;
const circleWrapperHeight = 54;
const circleWidth = 40;

export const deviceProps = {
  calloutBoxHeight,
  blueBoxHeight,
  blackBoxHeight,
  deviceVariableSizes,
  deviceScreen,
  defaultCenter,
  defaultZoom,
  deviceName,
  insideColPadding,
  circleWrapperHeight,
  circleWidth
};
