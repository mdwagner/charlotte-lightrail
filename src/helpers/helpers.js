import { Linking } from 'react-native';
import { mapboxApiKey } from './config';

export const startNavigation = async (mode, latlng) => {
  const { latitude, longitude } = latlng;

  const googleMapsURL = `comgooglemaps://?daddr=${latitude},${longitude}&directionsmode=${mode}`;
  const appleMapsURL = `http://maps.apple.com?daddr=${latitude},${longitude}&dirflg=${mode}`;

  const supported = await Linking.canOpenURL(googleMapsURL);
  if (supported) {
    Linking
      .openURL(googleMapsURL)
      .catch(err => console.error('Either Google Maps is not installed or some other error occurred: ', err));
  } else {
    Linking
      .openURL(appleMapsURL)
      .catch(err => console.error('An error occurred: ', err));
  }
};

export const mapboxDistanceAPI = {
  async getDistance(origin, destinations, mode) {
    const coordinates = destinations;
    coordinates.unshift(origin);

    if (__DEV__) {
      const endpoint = `directions-matrix/v1/mapbox/${mode}`;
      coordinates_string = coordinates.map(coordinate => {
        return coordinate.toString();
      });
      coordinates_query = coordinates_string.join(';');
      const url = `https://api.mapbox.com/${endpoint}/${coordinates_query}?access_token=${mapboxApiKey}`;
      const res = await fetch(url);
      return res.json();
    } else {
      const endpoint = `distances/v1/mapbox/${mode}`;
      const url = `https://api.mapbox.com/${endpoint}?access_token=${mapboxApiKey}`;
      const res = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          coordinates,
        }),
      });
      return res.json();
    }
  },
};