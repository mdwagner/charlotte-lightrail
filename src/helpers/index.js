export { withHelpers, withStaticData } from './hoc';