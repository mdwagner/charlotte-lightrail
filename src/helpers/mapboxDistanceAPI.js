import { mapboxApiKey } from './config';

export const mapboxDistanceAPI = {
  async getDistance(origin, destinations, mode) {
    const coordinates = destinations;
    coordinates.unshift(origin);

    if (__DEV__) {
      const endpoint = `directions-matrix/v1/mapbox/${mode}`;
      const coordinatesString = coordinates.map(coordinate => coordinate.toString());
      const coordinatesQuery = coordinatesString.join(';');
      const url = `https://api.mapbox.com/${endpoint}/${coordinatesQuery}?access_token=${mapboxApiKey}`;
      const response = await fetch(url);
      return response.json();
    }

    const endpoint = `distances/v1/mapbox/${mode}`;
    const url = `https://api.mapbox.com/${endpoint}?access_token=${mapboxApiKey}`;
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        coordinates
      })
    });
    return response.json();
  }
};
