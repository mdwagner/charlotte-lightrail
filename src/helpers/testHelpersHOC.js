import React from 'react';
import { ThemeProvider } from 'styled-components/native';
import renderer from 'react-test-renderer';
import { COLORS } from '../assets/styles/constants';

export const snapshotStyledComponent = (component) => {
  const snapshot = renderer.create(
    <ThemeProvider theme={COLORS}>
      {component}
    </ThemeProvider>
  ).toJSON();
  expect(snapshot).toMatchSnapshot();
};

export const themeStyledComponent = component => (
  <ThemeProvider theme={COLORS}>
    {component}
  </ThemeProvider>
);
