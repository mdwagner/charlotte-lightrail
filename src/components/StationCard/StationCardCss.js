import styled from 'styled-components/native';
import { deviceProps } from '../../helpers';

export const NearestContainerView = styled.View`
  height: ${deviceProps.deviceVariableSizes['calloutHeight']}px;
  width: ${deviceProps.deviceScreen.width}px;
`;

export const TriangleView = styled.View`
  flexDirection: row;
  alignSelf: center;
  width: 0px;
  height: 0px;
  backgroundColor: ${props => props.theme.transparent};
  borderStyle: solid;
  borderLeftWidth: 25px;
  borderRightWidth: 25px;
  borderBottomWidth: 25px;
  borderLeftColor: ${props => props.theme.transparent};
  borderRightColor: ${props => props.theme.transparent};
  borderBottomColor: ${props => props.theme.backgroundColorTrans};
`;

export const InfoContainerView = styled.View`
  height: ${deviceProps.blueBoxHeight};
  flexDirection: row;
  backgroundColor: ${props => props.theme.backgroundColorTrans};
  paddingHorizontal: 12px;
  paddingVertical: 0px;
`;

export const ArrowWrapperTouchableOpacity = styled.TouchableOpacity`
  justifyContent: center;
  padding: 6px;
`;

export const ArrowWrapperView = styled.View`
  justifyContent: center;
  padding: 6px;
`;

const ArrowView = styled.View`
  borderStyle: solid;
  borderTopWidth: 1px;
  borderRightWidth: 1px;
  borderBottomWidth: 0px;
  borderLeftWidth: 0px;
  borderColor: ${props => props.disabled ? props.theme.disabledArrow : props.theme.primaryTextColor};
  margin: 10px;
  height: 30px;
  width: 30px;
`;

export const LeftArrowView = ArrowView.extend`
  transform: rotate(-135deg);
`;

export const RightArrowView = ArrowView.extend`
  transform: rotate(45deg);
`;

export const StationLabelContainerView = styled.View`
  alignItems: center;
  justifyContent: center;
  flex: 1;
`;

export const StationLabelText = styled.Text`
  fontSize: ${deviceProps.deviceVariableSizes['stationLabelFontSize']}px;
  color: ${props => props.theme.primaryTextColor};
`;

export const SmallGrayText = styled.Text`
  color: ${props => props.theme.grayText};
  fontSize: 13px;
`;

export const TimesContainerView = styled.View`
  height: ${deviceProps.blackBoxHeight}px;
  flexDirection: row;
  alignItems: center;
  paddingHorizontal: 12px;
  backgroundColor: ${props => props.theme.backgroundColorDark};
`;

export const NextBlockView = styled.View`
  alignItems: center;
  flex: 1;
`;

export const NextTimeWrapView = styled.View`
  paddingVertical: ${deviceProps.deviceVariableSizes['nextTimeWrapPaddingVertical']}px;
`;

export const NextTimeText = styled.Text`
  fontSize: ${deviceProps.deviceVariableSizes['nextTimeFontSize']}px;
  color: ${props => props.theme.primaryTextColor};
`;

// left: Half device width - half of ((buttonWidth + buttonPadding) * numButtons)
export const ButtonContainerView = styled.View`
  position: absolute;
  height: 50px;
  bottom: ${deviceProps.blackBoxHeight - (50 / 2)}px;
  left: ${(deviceProps.deviceScreen['width'] / 2) - (((50 + 12) * 3) / 2)}px;
`;

export const ButtonsView = styled.View`
  flex: 1;
  flexDirection: row;
  alignSelf: center;
`;

export const BubbleTouchableOpacity = styled.TouchableOpacity`
  backgroundColor: ${props => props.theme.backgroundColorButton};
  borderRadius: 50px;
  borderWidth: 0px;
  marginHorizontal: 6px;
  padding: 7px;
  width: 50px;
  height: 50px;
`;

export const LoadingWrapperView = styled.View`
  backgroundColor: ${props => props.theme.backgroundColorTrans};
  height: ${deviceProps.calloutBoxHeight}px;
  justifyContent: center;
`;

export const LoadingText = styled.Text`
  textAlign: center;
  fontSize: 16px;
  paddingVertical: 10px;
  color: ${props => props.theme.grayText};
`;
