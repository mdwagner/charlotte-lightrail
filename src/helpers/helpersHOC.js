import React from 'react';
import * as configHelpers from './config';
import * as deviceHelpers from './device';
import * as helperHelpers from './helpers';
import * as mapSetupHelpers from './mapSetup';
import * as scheduleHelpers from './scheduleCalcs';

export function withHelpers(Component) {
  return class extends React.Component {
    render() {
      const newProps = {
        ...configHelpers,
        ...deviceHelpers,
        ...helperHelpers,
        ...mapSetupHelpers,
        ...scheduleHelpers
      };
      return (
        <Component {...this.props} helpers={newProps} />
      );
    }
  };
}
