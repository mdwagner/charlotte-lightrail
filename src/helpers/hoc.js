import React from 'react';
import * as configHelpers from './config';
import * as deviceHelpers from './device';
import * as helperHelpers from './helpers';
import * as mapSetupHelpers from './mapSetup';
import * as scheduleHelpers from './scheduleCalcs';
import staticData from './staticData.json';

export function withHelpers(Component) {
  return class extends React.Component {
    constructor(props) {
      super(props);
    }
    render() {
      const newProps = {
        ...configHelpers,
        ...deviceHelpers,
        ...helperHelpers,
        ...mapSetupHelpers,
        ...scheduleHelpers
      };
      return (
        <Component {...this.props} helpers={newProps} />
      );
    }
  };
}

export function withStaticData(Component) {
  return class extends React.Component {
    constructor(props) {
      super(props);
    }
    render() {
      const newProps = {
        staticData
      };
      return (
        <Component {...this.props} staticData={newProps} />
      );
    }
  };
}