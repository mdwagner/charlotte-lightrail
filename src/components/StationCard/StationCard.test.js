import React from 'react';
import { snapshotStyledComponent } from '../../helpers/testHelpersHOC';
import StationCard from './StationCard';

describe('<StationCard />', () => {
  let props;

  beforeAll(() => {
    props = {
      connected: true,
      loading: false,
      mode: 'walking',
      panToStation: jest.fn(),
      navigation: { navigate: jest.fn() }
    };
  });

  it('renders correctly', () => {
    snapshotStyledComponent(
      <StationCard {...props} />
    );
  });
});
