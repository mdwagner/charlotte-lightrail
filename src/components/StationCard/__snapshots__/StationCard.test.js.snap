exports[`<StationCard /> renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "height": 279,
        "width": 375,
      },
      undefined,
    ]
  }>
  <View
    style={
      Array [
        Object {
          "alignSelf": "center",
          "backgroundColor": "transparent",
          "borderBottomColor": "rgba(0, 20, 30, 0.8)",
          "borderBottomWidth": 25,
          "borderLeftColor": "transparent",
          "borderLeftWidth": 25,
          "borderRightColor": "transparent",
          "borderRightWidth": 25,
          "borderStyle": "solid",
          "flexDirection": "row",
          "height": 0,
          "width": 0,
        },
        undefined,
      ]
    } />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "rgba(0, 20, 30, 0.8)",
          "height": 254,
          "justifyContent": "center",
        },
        undefined,
      ]
    }>
    <ActivityIndicator
      animating={true}
      color="#999999"
      hidesWhenStopped={true}
      size="large" />
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#CCC",
            "fontSize": 16,
            "paddingVertical": 10,
            "textAlign": "center",
          },
          undefined,
        ]
      }>
      Finding nearest station...
    </Text>
  </View>
</View>
`;
